DC-1052: Improve getConceptHierarchy and getConceptChildren endpoint performance

snf2ye opened the pull request:
https://broadworkbench.atlassian.net/browse/DC-1052
_________________

In the original iteration of our getConceptHierarchy and getConceptChildren endpoints on Azure were taking more than 8 minutes to complete. This was longer than the timeout for our endpoints so these were ultimately unusable. In this PR, we when through two iterations of improving the query performance. 

### Iteration 1: Move to joins instead of subqueries in the select statement for the hasChildren computation 
In a previous spike, @pshapiro4broad had identified that calculating the "hasChildren" field on the getConceptHierarchy and getConceptChildren endpoints on Azure were taking the bulk of the time (over 90%). He found that rewriting the subqueries as a joins eliminated this overhead. The original goal of this ticket was to implement the rewritten queries. 

The original select statement
```
Select ... ,
  CASE WHEN EXISTS (SELECT 1 FROM concept_ancestor AS ca
     JOIN (SELECT * FROM concept AS c1
            ON c1.concept_id = ca.descendant_concept_id
         WHERE 
         (ca.ancestor_concept_id = cr.concept_id_2 
         AND ca.descendant_concept_id != cr.concept_id_2 
         AND c1.standard_concept = 'S')) 
  THEN 1 ELSE 0 END AS has_children
... 
FROM... 
```

Changing the code to use a LEFT JOIN looks like this:

```
SELECT count(ca.descendant_concept_id) AS has_children
  ...
         LEFT JOIN (SELECT ca.ancestor_concept_id,
                           ca.descendant_concept_id
                    FROM concept_ancestor AS ca
                             JOIN concept AS c1
                         ON
                             c1.concept_id = ca.descendant_concept_id
                    WHERE c1.standard_concept = 'S') AS ca2
                   ON
                       ca2.ancestor_concept_id = cr.concept_id_2
                           AND ca2.descendant_concept_id != cr.concept_id_2
```

For the getConceptHierarchy and for **most** concept ids with getConceptChildren, we saw huge performance gains with this re-written query. However, we saw some spikes for a few select concept ids with the getConceptChildren endpoint.

**Test case for the spike:**
Concept ID: 4042140 with the getConceptChildren endpoint
This spike was not unique to Azure. We were seeing the query timeout in Azure after 30 minutes and in BigQuery the query completed in 38 minutes. The execution graph in BigQuery pinpointed the join (the new left join to produce hasChildren). 
![image](https://github.com/DataBiosphere/jade-data-repo/assets/13254229/4167da99-2618-4947-864b-b95a49fc7325)


### Iteration 2: Narrow the scope of the new join by filtering on min_levels_of_separation
Looking the issue described in the BigQuery console, the issue was a high cardinality join. I think we were only seeing this for concepts that were at the top of the hierarchy, thus we were doing a join across many children. Walking down the children's relationships, we could see the time for query completion decrease at every generation. 

So, we went back to the drawing board.

**What exactly are we trying to calculate here?** We only want to find out if _any_ children exist for each descendent. We don't need to know about every descendent. 
**How could we limit the number of records we're joining on?** Only retrieve the first child of a concept instead of every child. We can do this by only grabbing the entries with a small  min_levels_of_separation. 

What is 'min_levels_of_separation'?  " a number of concepts between some ancestor_concept and descendant one in the hierarchy."

**One Catch**
Sometimes the min_levels_of_separation for the first child concept is 0 and sometimes it is 1. I don't know why or what is different between these two cases. But, after querying the concept_ancestor table, I feel confident that filtering to only include min_levels_of_separation <= 1 should account for all concepts that have children. The minimum min_levels_of_separation is always 0 or 1. 

![image](https://github.com/DataBiosphere/jade-data-repo/assets/13254229/ada6bc43-7389-42e6-9758-ba7f3a00dd7b)


### The Results

![Improving Query Performance_ getConceptHierarchy (4)](https://github.com/DataBiosphere/jade-data-repo/assets/13254229/44eef701-72c9-4800-9ec7-142c1cd43f57)

![Improving Query Performance_ getConceptChildren (1)](https://github.com/DataBiosphere/jade-data-repo/assets/13254229/757ab8c3-7bb2-44d7-b747-f2dfddf9bd37)


From profiling the endpoints, we can see that iteration one significantly improved performance for most concepts tested except when we saw a spike in the query time. We saw more consistent but slightly slower results in iteration 2. Iteration 2 should at least allow the endpoints to complete without timing out. 

**Cavet** 
While testing, I found that we could suddenly run into a spike in performance for synapse. Possibly another job was running against synapse causing it to slow down? 
![Screenshot 2024-05-22 at 4 58 00 PM](https://github.com/DataBiosphere/jade-data-repo/assets/13254229/098caedb-2c40-40b5-9f2e-b3a24f9defe5)

**Update** Profiling these endpoints on a new, dedicated synapse instance provided more consistent results without spikes.

snf2ye committed:
Share TableVariable/TablePointer join logic for subquery

- Rename TableVariable -> SourceVariable
- Add SourcePointer interface and have TablePointer implement the interface (in the next commit, I'll add a SubQueryPointer class to also implements the SourcePointer interface)
- Move all references to TablePointer -> SourcePointer
- Add method to SourcePointer to retrieve source naming method to be used in aliasing

snf2ye committed:
Create SubQueryPointer.java

snf2ye committed:
Update HierarchyQueryBuilder to use a left join instead of the subquery for has_children

snf2ye committed:
Handle multiple join clauses

snf2ye committed:
Working GCP hierarchy query

snf2ye committed:
Working Azure hierarchy query

We only need the '>0' clause for gcp

remove for azure query

Fix unit tests

snf2ye committed:
Concept Children query: Move where clause to join

snf2ye committed:
Improve query performance by joining on min_levels_of_separation = 1 for the has_children clause

snf2ye committed:
profile endpoints script updates

- Write results to a gitignored file
- pull concept ids from a csv file

s-rubenstein commented on src/main/java/bio/terra/service/snapshotbuilder/query/FieldVariable.java:
```diff
@@ -70,6 +70,10 @@ public String renderSQL(SqlRenderContext context) {
     }
 
+    if (fieldPointer.hasHasCountClause() && context.getPlatform().isGcp()) {
+      sql = sql + " " + fieldPointer.getHasCountClause();
```
I think you've explained this already, but is there any special code for azure count clause?

snf2ye replied:
I believe that azure expects a numeric value to set the boolean (if 0, false, otherwise true). For GCP, we need to convert the count into a boolean. So, this "hasCountClause" is doing a comparison for GCP to determine if the count > 0, return true, otherwise false. 

This implementation here is a bit wonky so might be worth reworking this code to generate the special case for gcp.

rjohanek replied:
if you wanted you could use the platform choose method here, although if you're doing nothing for the azure case maybe it doesn't make sense to

pshapiro4broad commented on src/test/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilderTest.java:
```diff
@@ -147,10 +147,10 @@ SELECT c.concept_name, c.concept_id, c.concept_code, COUNT(DISTINCT co.person_id
   void testCreateSearchConceptClause(SqlRenderContext context) {
-    TableVariable conceptTableVariable =
-        TableVariable.forPrimary(TablePointer.fromTableName(Concept.TABLE_NAME));
+    SourceVariable conceptSourceVariable =
```
To me, this shows why it's not necessarily good to name variables after their types. In this code's context, is the type more important than the thing it represents? Maybe now is a good time to rename it to `conceptTable` so it refers to its content instead of its type.
```suggestion
    SourceVariable conceptTable =
```

s-rubenstein commented on src/main/java/bio/terra/service/snapshotbuilder/query/SourceVariable.java:
```diff
@@ -0,0 +1,118 @@
+    String alias = context.getAlias(this);
+
+    if (alias != null) {
+      sql = new ST("<sql> AS <tableAlias>").add("sql", sql).add("tableAlias", alias).render();
```
Do you think this should be just `alias` rather than `tableAlias` now? Since it could be from places other than tables?

s-rubenstein commented on src/main/java/bio/terra/service/snapshotbuilder/utils/HierarchyQueryBuilder.java:
```diff
@@ -92,73 +84,68 @@ public Query generateQuery(SnapshotBuilderDomainOption domainOption, int concept
-  static SelectExpression hasChildrenExpression(FieldVariable conceptId) {
-    var conceptAncestor =
-        TableVariable.forPrimary(TablePointer.fromTableName(ConceptAncestor.TABLE_NAME));
+  static FieldVariable hasChildrenSelect(SourceVariable hasChildrenJoin) {
```
How do you feel about `makeChildrenSelect` or something similar to that? `has` to me implies a boolean response type, and while I know the IDE will help us out, I always like to have the name of the method imply things for us.

snf2ye replied:
Ah, yeah - I could do "makeHasChildrenSelect".

s-rubenstein commented on src/main/java/bio/terra/service/snapshotbuilder/utils/HierarchyQueryBuilder.java:
```diff
@@ -92,73 +84,68 @@ public Query generateQuery(SnapshotBuilderDomainOption domainOption, int concept
+  }
+
+  /** Generate a join clause that is used to determine if the outerConcept has any children. */
+  static SourceVariable hasChildrenJoin(FieldVariable childId) {
```
(Same comment about `make` vs `has`)

s-rubenstein commented on tools/profileEndpoints/concept_ids.csv:
Should this file be excluded?

snf2ye replied:
I pulled the test concept_id out of the profile script and into this file, so it's needed to run the script.

s-rubenstein commented on tools/profileEndpoints/profile_endpoints.py:
```diff
@@ -261,7 +271,7 @@ def profile_enumerate_concepts(token):
         else:
             url = f"concepts?domainId={domain_id}&filterText={filter_text}"
-        make_get_request(url, token)
+        make_get_request(url, token,1)
```
Nit: Missing space

snf2ye committed:
Fix rollup count and hasChildren

- There is a case where a child concept only has one entry in the concept_ancestor table and its minimum separation is 0 instead of 1 (Not sure why). So, switching to have min_levels_of_separation <= 1 instead of just equal to one should encompass all of the cases.
- With my previous changes, most of the rollup counts for the hierarchy endpoint were returning as zero.  A more complete join with the concept_ancestor table fixes this but probably impacts performance.
- Fix hierarchy tests by adding min_levels_of_separation to concept_ancestor test data

snf2ye committed:
Code Review changes

rjohanek commented on src/main/java/bio/terra/service/snapshotbuilder/utils/constants/ConceptAncestor.java:
```diff
@@ -6,4 +6,5 @@ private ConceptAncestor() {}
   public static final String TABLE_NAME = "concept_ancestor";
   public static final String ANCESTOR_CONCEPT_ID = "ancestor_concept_id";
   public static final String DESCENDANT_CONCEPT_ID = "descendant_concept_id";
+  public static final String MIN_LEVELS_OF_SEPARATION = "min_levels_of_separation";
```
Do we still need this?

snf2ye replied:
Yes, it is still used to calculate has_children (As a filter so that we don't have to join on so many records)

rjohanek commented on src/test/resources/omop/concept-ancestor-table-data.jsonl:
```diff
@@ -1,10 +1,10 @@
-{"descendant_concept_id": 4, "ancestor_concept_id": 3}
-{"descendant_concept_id": 4, "ancestor_concept_id": 4}
-{"descendant_concept_id": 5, "ancestor_concept_id": 5}
+{"descendant_concept_id": 1, "ancestor_concept_id": 2, "min_levels_of_separation": 0}
```
If we aren't going to use min levels of separation anymore, can these changes be reverted?

rjohanek commented on src/main/java/bio/terra/service/snapshotbuilder/utils/HierarchyQueryBuilder.java:
```diff
@@ -76,10 +77,16 @@ public Query generateQuery(SnapshotBuilderDomainOption domainOption, int concept
             count,
-            hasChildrenExpression(childId)),
-        List.of(conceptRelationship, child, parent, conceptAncestor, domainOccurrence),
+            selectHChildren(joinHasChildren)),
```
selectHasChildren?

snf2ye replied:
Gah, thank you. Fixed!

rjohanek commented on src/main/java/bio/terra/service/snapshotbuilder/utils/HierarchyQueryBuilder.java:
```diff
@@ -92,73 +99,69 @@ public Query generateQuery(SnapshotBuilderDomainOption domainOption, int concept
-        TableVariable.forPrimary(TablePointer.fromTableName(ConceptAncestor.TABLE_NAME));
+  static FieldVariable selectHChildren(SourceVariable joinHasChildren) {
+    return joinHasChildren.makeFieldVariable(
+        ConceptAncestor.DESCENDANT_CONCEPT_ID, "COUNT", "has_children", "> 0", true);
```
Does this work on azure as well?

snf2ye replied:
Good question - The "> 0" clause is only applied to the GCP version.

pshapiro4broad replied:
The way this column is interpreted by our code is under our control. So instead of generating different SQL for gcp and azure here, we could also change the code that reads the SQL response.

For example in `AggregateBQQueryResultsUtils`:
```
        .hasChildren(row.get(QueryBuilderFactory.HAS_CHILDREN).getLongValue() > 0)
```

rjohanek commented on src/main/java/bio/terra/service/snapshotbuilder/utils/HierarchyQueryBuilder.java:
```diff
@@ -92,73 +99,69 @@ public Query generateQuery(SnapshotBuilderDomainOption domainOption, int concept
-        TableVariable.forJoined(
+        conceptAncestorTable.makeFieldVariable(ConceptAncestor.DESCENDANT_CONCEPT_ID);
+    var minLevelsSeparation =
+        conceptAncestorTable.makeFieldVariable(ConceptAncestor.MIN_LEVELS_OF_SEPARATION);
```
If there are any children, at any level, don't we want to include them for the purposes of hasChildren? Does the min level of separation matter?

rjohanek replied:
Oh this is an optimization so we don;t need to compute such large values. Thanks for explaining that!

rjohanek commented on src/main/java/bio/terra/service/snapshotbuilder/utils/ConceptChildrenQueryBuilder.java:
```diff
@@ -34,35 +38,35 @@ public class ConceptChildrenQueryBuilder {
             TablePointer.fromTableName(CONCEPT_ANCESTOR), ANCESTOR_CONCEPT_ID, conceptId);
     FieldVariable descendantConceptId = conceptAncestor.makeFieldVariable(DESCENDANT_CONCEPT_ID);
 
+    var subQuery = createSubQuery(parentConceptId);
```
maybe there's a more specific name for this like selectChildren or something

rjohanek commented on src/main/java/bio/terra/service/snapshotbuilder/utils/ConceptChildrenQueryBuilder.java:
```diff
@@ -34,35 +38,35 @@ public class ConceptChildrenQueryBuilder {
     FieldVariable descendantConceptId = conceptAncestor.makeFieldVariable(DESCENDANT_CONCEPT_ID);
 
+    var subQuery = createSubQuery(parentConceptId);
+    var subQueryPointer = new SubQueryPointer(subQuery, "join_filter");
```
Maybe theres a clearer name for the subquery as well instead of "join_filter" just to help make this more readable

rjohanek commented on src/main/java/bio/terra/service/snapshotbuilder/utils/HierarchyQueryBuilder.java:
```diff
@@ -92,73 +99,69 @@ public Query generateQuery(SnapshotBuilderDomainOption domainOption, int concept
-                conceptAncestor.makeFieldVariable(ConceptAncestor.DESCENDANT_CONCEPT_ID),
-                conceptIdLiteral),
-            BinaryFilterVariable.notEquals(ancestorConceptId, conceptIdLiteral)));
+  static SourceVariable joinToFilterConcepts(FieldVariable parentId, int conceptId) {
```
Maybe rename to something like joinParents (or joinChildren, I see the id passed in is a parent) - or what does this method do exactly? Maybe a test for this would be beneficial for terms of documentation

snf2ye replied:
Great point - I've added comments and tests for all of these methods.

rjohanek commented on src/main/java/bio/terra/service/snapshotbuilder/utils/ConceptChildrenQueryBuilder.java:
```diff
@@ -104,8 +101,8 @@ public Query buildConceptChildrenQuery(
    */
   Query createSubQuery(int conceptId) {
```
maybe we could add a test for this as well

rjohanek reviewed (approved):
Looks good overall, just some comments on testing and renaming

s-rubenstein reviewed (approved):
LGTM pending the change Rae requested

snf2ye committed:
Typo fix

snf2ye committed:
Rename

snf2ye committed:
Add tests and comments

spotless
Remove result files

snf2ye committed:
Add tests and comments

spotless

pshapiro4broad commented on tools/profileEndpoints/performance-synapse/results_2024-05-23_14-51-28.csv:
```diff
@@ -0,0 +1,12 @@
+http://localhost:8080/api/repository/v1/snapshots/c3eb4708-444f-4cbf-a32c-0d3bb93d4819/snapshotBuilder/concepts/443883/hierarchy,443883,83.82 seconds
```
I think these got added by accident, before the `.gitignore` was updated

snf2ye replied:
Oh shoot, yes, I did not mean to check these in.

pshapiro4broad committed:
Change gcp has_children to return an int value like azure does, to make the query generation code simpler. Also remove some test result files

pshapiro4broad committed:
Merge remote-tracking branch 'origin/sh/dc-1052-subquery' into sh/dc-1052-subquery

# Conflicts:
#	src/main/java/bio/terra/service/snapshotbuilder/query/FieldVariable.java

pshapiro4broad committed:
merge conflicts
